index.js

```diff
@@ -55,21 +55,56 @@ function findAllDependencies(file, knownDependencies, baseDir, knownFiles) {
     return knownDependencies;
   }
 
-  if (baseDir) {
-    return findAllDependenciesHelp(file, knownDependencies, baseDir, knownFiles).then(function(thing){
+  if (sourceDirectories) {
+    return findAllDependenciesHelp(file, knownDependencies, sourceDirectories, knownFiles).then(function(thing){
       return thing.knownDependencies;
     });
   } else {
-    return getBaseDir(file).then(function(newBaseDir) {
-      return findAllDependenciesHelp(file, knownDependencies, newBaseDir, knownFiles).then(function(thing){
-        return thing.knownDependencies;
+    return getBaseDir(file)
+      .then(getElmPackageSourceDirectories)
+      .then(function(newSourceDirs) {
+        return findAllDependenciesHelp(file, knownDependencies, newSourceDirs, knownFiles).then(function(thing){
+          return thing.knownDependencies;
+        });
       });
-    })
   }
 }
 
+// Given a source directory (containing top-level Elm modules), locate the
+// elm-package.json file that includes it and get all its source directories.
+function getElmPackageSourceDirectories(baseDir, currentDir) {
+  if (typeof currentDir === "undefined") {
+    currentDir = baseDir = path.resolve(baseDir);
+  }
+
+  var elmPackagePath = path.join(currentDir, 'elm-package.json');
+  if (fs.existsSync(elmPackagePath)) {
+    var sourceDirectories = getSourceDirectories(elmPackagePath);
+    if (_.includes(sourceDirectories, baseDir)) {
+      return sourceDirectories;
+    }
+  }
+
+  if (isRoot(currentDir)) {
+    return false;
+  }
+
+  return getElmPackageSourceDirectories(baseDir, path.dirname(currentDir));
+}
+
+function isRoot(dir) {
+  var parsedPath = path.parse(dir);
+  return parsedPath.root === parsedPath.dir;
+}
 
-function findAllDependenciesHelp(file, knownDependencies, baseDir, knownFiles) {
+function getSourceDirectories(elmPackagePath) {
+  var elmPackage = JSON.parse(fs.readFileSync(elmPackagePath, 'utf8'));
```

Comment:
this will crash if there is invalid json, with a bad error message. Good to catch it and return an empty list instead.